inc/azure_c_shared_utility/threadapi.h

```diff
@@ -45,7 +45,7 @@ typedef void* THREAD_HANDLE;
  * @return	@c THREADAPI_OK if the API call is successful or an error
  * 			code in case it fails.
  */
-MOCKABLE_FUNCTION(, THREADAPI_RESULT, ThreadAPI_Create, THREAD_HANDLE*, threadHandle, THREAD_START_FUNC, func, void*, arg);
+DLLEXPORT MOCKABLE_FUNCTION(, THREADAPI_RESULT, ThreadAPI_Create, THREAD_HANDLE*, threadHandle, THREAD_START_FUNC, func, void*, arg);
```

Comment:
>DLLEXPORT [](start = 0, length = 9)

This relies of the fact that MOCKABLE_FUNCTION must generate a function as the first thing it does. Putting this assumtion in is a failure waiting to happen if MOCKABLE_FUNCTION changes slightly.
Can you just put DLLEXPORT as the first argument in the MOCKABLE_FUNCTION macro?

Reply:
No. I tried that. The issue is that the MOCKABLE_FUNCTION macro generates (in this example) THREADAPI_RESULT modifier ThreadAPI_Create(args...). In some instances in the SDK we already have the value WINAPI passed into the MOCKABLE_FUNCTION macro which works fine because __stdcall _must_ be declared after the return type. However, __declspec(dllexport) needs to be declared _before_ the return type so it needs to read __declspec(dllexport) THREADAPI_RESULT ThreadAPI_Create. I toyed with the idea of a premodifier and a postmodifier but really didn't want to mess about with the MOCKABLE _FUNCTION macro. 

I am open to ideas...

---
In reply to: [184155255](https://github.com/Azure/azure-c-shared-utility/pull/152#discussion_r184155255) [](ancestors = 184155255)